lms/services/event.py

```diff
@@ -0,0 +1,33 @@
+from functools import lru_cache
+
+from sqlalchemy.orm import Session
+
+from lms.models import Event, EventData, EventType, EventUser
+
+
+class EventService:
+    def __init__(self, db: Session):
+        self._db = db
+
+    def insert_event(
+        self, type_: EventType, user: None, lti_roles=None, data=None, **kwargs
+    ):
+
+        event = Event(type_id=self._get_type_pk(type_), **kwargs)
+        self._db.add(event)
+
+        if user:
+            for lti_role in lti_roles or [None]:
+                self._db.add(EventUser(event=event, user=user, lti_role=lti_role))
```

Comment:
We are not upserting here, just inserting. SQLA will generate one INSERT statement per role but it will send them all at once to the DB.

We can use the ORM, no need to optimize here.